Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -58,36 +84,248 @@ polynomials of elliptic curves, omitting a factor of the bivariate `2`-division
+    Int.mul_ediv_cancel' <| Int.dvd_add (h <| parity.2.1 ▸ parity.1) <| h parity.2.2]
+  ring_nf
+
+@[simp]
+lemma map_atomRel (W : ℤ → R) (p q r s : ℤ) : f (atomRel W p q r s) = atomRel (f ∘ W) p q r s := by
+  simp_rw [atomRel, map_add, map_sub, map_mul, map_atom]
+
+/-- The elliptic relator `ER(p, q, r, s)` that defines an elliptic net. -/
+def rel (p q r s : ℤ) : R :=
+  W (p + q + s) * W (p - q) * W (r + s) * W r - W (p + r + s) * W (p - r) * W (q + s) * W q +
+    W (q + r + s) * W (q - r) * W (p + s) * W p
+
+lemma rel_eq (p q r s : ℤ) : rel W p q r s = atomRel W (2 * p + s) (2 * q + s) (2 * r + s) s := by
+  simp_rw [rel, atomRel, atom, add_add_add_comm _ s, add_assoc _ s, ← two_mul, ← mul_add,
+    add_sub_add_comm, add_sub_assoc, sub_self, add_zero, ← mul_sub,
+    Int.mul_tdiv_cancel_left _ two_ne_zero, mul_comm <| _ * W p, mul_assoc]
 
-section IsEllDivSequence
+lemma atomRel_two_mul (p q r s : ℤ) :
+    atomRel W (2 * p) (2 * q) (2 * r) (2 * s) = rel W (p - s) (q - s) (r - s) (2 * s) := by
+  simp_rw [rel_eq, mul_sub, sub_add_cancel]
 
-variable (W : ℤ → R)
+lemma atomRel_eq {p q r s : ℤ} (parity : s % 2 = p % 2 ∧ s % 2 = q % 2 ∧ s % 2 = r % 2) :
+    atomRel W p q r s = rel W ((p - s) / 2) ((q - s) / 2) ((r - s) / 2) s := by
+  simp only [rel_eq, Int.mul_ediv_cancel', Int.ModEq.dvd parity.1, Int.ModEq.dvd parity.2.1,
+    Int.ModEq.dvd parity.2.2, sub_add_cancel]
 
-/-- The proposition that a sequence indexed by integers is an elliptic sequence. -/
+variable {W} in
+@[simp]
+lemma rel_neg (odd : ∀ n : ℤ, W (-n) = -W n) (p q r s : ℤ) :
+    rel W (-p) (-q) (-r) (-s) = rel W p q r s := by
+  simp_rw [rel_eq, mul_neg, ← neg_add, atomRel_neg₁ odd, atomRel_neg₂ odd, atomRel_neg₃ odd,
+    atomRel_neg₄]
+
+lemma rel_even (m : ℤ) : rel W (m + 1) (m - 1) 1 0 = W (2 * m) * W 2 * W 1 ^ 2 -
+    W (m - 1) ^ 2 * W m * W (m + 2) + W (m - 2) * W m * W (m + 1) ^ 2 := by
+  rw [rel]
+  ring_nf
+
+lemma rel_odd (m : ℤ) : rel W (m + 1) m 1 0 =
+    W (2 * m + 1) * W 1 ^ 3 - W (m + 2) * W m ^ 3 + W (m - 1) * W (m + 1) ^ 3 := by
+  rw [rel]
+  ring_nf
+
+@[simp]
+lemma map_rel (W : ℤ → R) (p q r s : ℤ) : f (rel W p q r s) = rel (f ∘ W) p q r s := by
+  simp_rw [rel, map_add, map_sub, map_mul, Function.comp]
+
+end IsEllipticNet
+
+/-- The proposition that a sequence indexed by `ℤ` is an elliptic net. -/
+def IsEllipticNet : Prop :=
+  ∀ p q r s : ℤ, IsEllipticNet.rel W p q r s = 0
+
+/-- The proposition that a sequence indexed by `ℤ` is an elliptic sequence. -/
 def IsEllSequence : Prop :=
-  ∀ m n r : ℤ, W (m + n) * W (m - n) * W r ^ 2 =
-    W (m + r) * W (m - r) * W n ^ 2 - W (n + r) * W (n - r) * W m ^ 2
+  ∀ p q r : ℤ, IsEllipticNet.rel W p q r 0 = 0
 
-/-- The proposition that a sequence indexed by integers is a divisibility sequence. -/
+/-- The proposition that a sequence indexed by `ℤ` is a divisibility sequence. -/
 def IsDivSequence : Prop :=
   ∀ m n : ℕ, m ∣ n → W m ∣ W n
```

Comment:
This does not impose any constraints on the terms with negative indices. Is this intentional?